textattack/transformations/white_box/gradient_based_word_swap.py

```diff
@@ -0,0 +1,152 @@
+from textattack.transformations import Transformation
+
+class GradientBasedWordSwap(Transformation):
+    """ Uses the model's gradient to suggest replacements for a given word.
+        
+        Based off of HotFlip: White-Box Adversarial Examples for Text 
+            Classification (Ebrahimi et al., 2018).
+        
+            https://arxiv.org/pdf/1712.06751.pdf
+        
+        Arguments:
+            model (nn.Module): The model to attack. Model must have a 
+                `word_embeddings` matrix and `convert_id_to_word` function.
+            top_n (int): the number of top words to return at each index
+            replace_stopwords (bool): whether or not to replace stopwords
+    """
+    def __init__(self, model, top_n=1, replace_stopwords=False):
+        validate_model_gradient_word_swap_compatibility(model)
+        if not hasattr(model, 'word_embeddings'):
+            raise ValueError('Model needs word embedding matrix for gradient-based word swap')
+        if not hasattr(model, 'lookup_table'):
+            raise ValueError('Model needs lookup table for gradient-based word swap')
+        if not hasattr(model, 'zero_grad'):
+            raise ValueError('Model needs `zero_grad()` for gradient-based word swap')
+        if not hasattr(model.tokenizer, 'convert_id_to_word'):
+            raise ValueError('Tokenizer needs `convert_id_to_word()` for gradient-based word swap')
+        if not hasattr(model.tokenizer, 'pad_id'):
+            raise ValueError('Tokenizer needs `pad_id` for gradient-based word swap')
+        if not hasattr(model.tokenizer, 'oov_id'):
+            raise ValueError('Tokenizer needs `oov_id` for gradient-based word swap')
+        self.loss = torch.nn.CrossEntropyLoss()
+        self.model = model
+        self.pad_id = self.model.tokenizer.pad_id
+        self.oov_id = self.model.tokenizer.oov_id
+        self.top_n = top_n
+        if replace_stopwords:
+            self.stopwords = set()
+        else:
+            from nltk.corpus import stopwords
+            self.stopwords = set(stopwords.words('english'))
+
+    def _get_replacement_words_by_grad(self, text, indices_to_replace):
+        """ Returns returns a list containing all possible words to replace
+            `word` with, based off of the model's gradient.
+            
+            Arguments:
+                text (TokenizedText): The full text input to perturb
+                word_index (int): index of the word to replace
+        """
+        self.model.train()
+       
+        lookup_table = self.model.lookup_table.to(utils.get_device())
+        lookup_table_transpose = lookup_table.transpose(0,1)
+        
+        # set backward hook on the word embeddings for input x
+        emb_hook = Hook(self.model.word_embeddings, backward=True)
+    
+        self.model.zero_grad()
+        predictions = self._call_model(text)
+        original_label = predictions.argmax()
```

Comment:
I realized also that the gradient with respect to a single input isn't defined for models that take multiple inputs – like entailment models – so that's why it has to look like this